_posts/2022-09-14-utilizing-state-machines-in-front-end-development.md

```diff
@@ -0,0 +1,543 @@
+
+By growing the app we are starting to encounter some corner cases that we need to think about. Could the `isLoading` and `hasError` boolean flags be both true at the same time? How about the new `hasSearched` flag? Which combinations of these three are possible and what should we print out for each combination? **Every boolean flag we are adding multiplies the possible combinations, and every combination should be handled one way or another.** Every bit of state makes the component more complicated and more difficult to comprehend.
+
+So, how could we fix this? If you are thinking of state machines, you are correct!
+
+# State machines to the rescue
+
+First, let's define the states the component can be in. We'll use a plain object for easier autocompletion and typo-proofing. If we were using TypeScript, a union type or an enum would work even better.
+
+```js
+const states = {
+  INITIAL: "INITIAL",
+  LOADING: "LOADING",
+  FAILURE: "FAILURE",
+  NO_RESULTS: "NO_RESULTS",
+  SHOW_RESULTS: "SHOW_RESULTS",
+};
+```
+
+Now we can use just one variable for the current state and it will always be set to one of the predefined values. Generating the output is now much easier since we know all the possible states the component can be in. No more dozens of combinations or corner cases.
+
+```jsx
+const SearchForm = () => {
+  const [state, setState] = useState(states.INITIAL);
+  const [keyword, setKeyword] = useState("");
+  const [results, setResults] = useState([]);
+
+  const search = (event) => {
+    event.preventDefault();
+
+    setState(states.LOADING);
+
+    fetchResults(keyword)
+      .then((results) => {
+        setResults(results);
+        setState(results.length > 0 ? states.SHOW_RESULTS : states.NO_RESULTS);
+      })
+      .catch((error) => setState(states.FAILURE));
+  };
+
+  return (
+    <div>
+      <form onSubmit={search}>
+        <input onChange={(event) => setKeyword(event.target.value)} />
+        <button type="submit" disabled={state === states.LOADING}>
+          Search
+        </button>
+      </form>
+
+      {state === states.INITIAL && <p>Hi! Enter keyword please.</p>}
+      {state === states.LOADING && <p>Searching...</p>}
+      {state === states.FAILURE && <p>Oh no!</p>}
+      {state === states.NO_RESULTS && <p>Sorry! No results.</p>}
+      {state === states.SHOW_RESULTS &&
+        results.map((result) => <p>{result}</p>)}
+    </div>
+  );
+};
+```
+
+Much better already. But it's not quite a proper state machine yet. Referring to the definition of an FSM at the start of the blog post, we know that now our component can be only in one state of a finite set. We also transition to different states based on inputs. Our initial state is set. But we haven't defined the conditions for each transition yet.
+
+We could fix that by defining a list of possible transitions. While at it, let's add a utility function for checking if a transition from state `A` to `B` is valid.
```

Comment:
Here I thought - what if the transition is invalid? Should we end up in the error state rather than just ignore the invalid transition? It's good to to keep things simple so maybe this is irrelevant in this blog. Just wondering...

Reply:
When using XState, you can always send the events, and the machine decides what it does with them. To mimic that and to keep the code as short as possible, I'll leave it as it is. Good point tho.